azurelinuxagent/ga/cpucontroller.py

```diff
@@ -0,0 +1,269 @@
+
+    def _get_cpu_ticks(self, allow_no_such_file_or_directory_error=False):
+        """
+        Returns the number of USER_HZ of CPU time (user and system) consumed by this cgroup.
+
+        If allow_no_such_file_or_directory_error is set to True and cpuacct.stat does not exist the function
+        returns 0; this is useful when the function can be called before the cgroup has been created.
+        """
+        try:
+            cpuacct_stat = self._get_file_contents('cpuacct.stat')
+        except Exception as e:
+            if not isinstance(e, (IOError, OSError)) or e.errno != errno.ENOENT:  # pylint: disable=E1101
+                raise CGroupsException("Failed to read cpuacct.stat: {0}".format(ustr(e)))
+            if not allow_no_such_file_or_directory_error:
+                raise e
+            cpuacct_stat = None
+
+        cpu_ticks = 0
+
+        if cpuacct_stat is not None:
+            #
+            # Sample file:
+            #     # cat /sys/fs/cgroup/cpuacct/azure.slice/walinuxagent.service/cpuacct.stat
+            #     user 10190
+            #     system 3160
+            #
+            match = re_v1_user_system_times.match(cpuacct_stat)
+            if not match:
+                raise CGroupsException("The contents of {0} are invalid: {1}".format(self._get_cgroup_file('cpuacct.stat'), cpuacct_stat))
+            cpu_ticks = int(match.groups()[0]) + int(match.groups()[1])
+
+        return cpu_ticks
+
+    def get_cpu_usage(self):
+        if not self._cpu_usage_initialized():
+            raise CGroupsException("initialize_cpu_usage() must be invoked before the first call to get_cpu_usage()")
+
+        self._previous_cgroup_cpu = self._current_cgroup_cpu
+        self._previous_system_cpu = self._current_system_cpu
+        self._current_cgroup_cpu = self._get_cpu_ticks()
+        self._current_system_cpu = self._osutil.get_total_cpu_ticks_since_boot()
+
+        cgroup_delta = self._current_cgroup_cpu - self._previous_cgroup_cpu
+        system_delta = max(1, self._current_system_cpu - self._previous_system_cpu)
+
+        return round(100.0 * self._osutil.get_processor_cores() * float(cgroup_delta) / float(system_delta), 3)
+
+    def get_cpu_throttled_time(self, read_previous_throttled_time=True):
+        # Throttled time is reported in nanoseconds in v1
+        if not read_previous_throttled_time:
+            return float(self._get_cpu_stat_counter(counter_name='throttled_time') / 1E9)
+
+        if not self._cpu_usage_initialized():
+            raise CGroupsException("initialize_cpu_usage() must be invoked before the first call to get_cpu_throttled_time()")
+
+        self._previous_throttled_time = self._current_throttled_time
+        self._current_throttled_time = self._get_cpu_stat_counter(counter_name='throttled_time')
+
+        return round(float(self._current_throttled_time - self._previous_throttled_time) / 1E9, 3)
+
+
+class CpuControllerV2(_CpuController):
+    def initialize_cpu_usage(self):
```

Comment:
may be this can moved to abstract class?

Reply:
I started to make this change and I think it made the code more complicated. I would have to create abstractions for all of these -> _get_cpu_time/_get_cpu_ticks _get_system_usage/get_total_cpu_ticks_since_boot, _get_cpu_stat_counter(counter_name='throttled_time')/_get_cpu_stat_counter(counter_name='throttled_usec')